### ISSUE TITLE
Is importing Google Timeline supposed to take more than 2 weeks?

### ISSUE THREAD

**Issue by hopeseekr** [Freika/dawarich#500]
**Describe the bug**
I started the import of Google Timeline history since 2014 on 30 November 

**Version**
0.8.2

**To Reproduce**
1. Follow the `docker cp` steps to manually import a 754 MB Google Timeline Records.json.
2. Wait an indefinite amount of time.
3. Look at https://location.zpf.io/sidekiq/?days=7 
4. It's processing exceptionally slowly. 
5. After 6 days, it's only indexed 2014 partial 2016.

**Expected behavior**
I'd expect to to process more than 25 requests/second. If there are 3 million requests, it claims it would process them all in 1.48 days, but it's taking substantially longer.

It seems to be doing about 290,000 look ups a day (~12000 per hour), or 3.35 look ups per second.

With 1.786.000 more requests, that seems to be an additional 6 days 3 hours 48 minutes... So that would mean total time would be around 12.5 days.

**Screenshots**
![image](https://github.com/user-attachments/assets/8a32daed-f767-46ab-9563-9f228ce03f8d)

**Logs**
![image](https://github.com/user-attachments/assets/ee90496b-b752-4b96-ac40-00b86989ad93)

**Additional context**
The server is really beefy.

128 GB of RAM, 49.8 GB 100% free, over 70 GB available.
Enterprise NVME drives running on RAID-1.

Top of the line  Intel Xeon E-2176G CPU @ 3.70GHz, 12 cores.

```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                              
3217656 root      20   0  621336 136076  14720 S  50.9   0.2   6:00.15 bundle  
```

The `bundle` app seems stuck at 50% CPU on 1 core...

**Comment by hopeseekr** [comment]
I would like to contribute a Records.json parser in Rust to this project.  Or if you prefer, I'll host it separately. Your call.

Could you please give me some details on what part  of Records.json is consumed and goes into which tables?

**Comment by bklaesener** [comment]
>   The bundle app seems stuck at 50% CPU on 1 core...

Are you aware of this section of docker-compose.yml?

```
deploy:
 resources:
   limits:
     cpus: '0.50'    # Limit CPU usage to 50% of one core
     memory: '2G'    # Limit memory usage to 2GB` 
```

**Comment by Freika** [comment]
@hopeseekr Import takes couple hours maximum, what takes that long is reverse geocoding, in case you're using default Nominatim API. It limits number of requests to 1 per second.

With that in mind, I appreciate your offer to provide importer written in a language that is faster than Ruby, I'm not sure if it really necessary.

This guide should help you with speeding up your reverse geocoding process: https://dawarich.app/docs/tutorials/reverse-geocoding#setting-up-your-own-reverse-geocoding-service

**Comment by Beer17HWAM** [comment]
How van you tell us that it's using Nominatim?? In the docker-compose.yml file I see:

PHOTON_API_HOST: photon.komoot.io
PHOTON_API_USE_HTTPS: true

And that is also very slow!!

I want to setup my own-reverse-geocoding-service with photon locally in a docker like you told us. Is it also that slow???

**Comment by Freika** [comment]
If the host is photon.komoot.io, then it's being rate-limited by Dawarich to not exceed Nominatim's limits. If you provide different url, no rate limits applied by Dawarich.

**Comment by hopeseekr** [comment]
Those of us with cognitive difficulties who use Google Timeline Desktop as a metacortex to remember where we were 3 days ago... I personally am having a big problem with November before I got this all running on Nov 30th.... 

The parser really should do a LIFO approach. I'm really struggling. Fortunately, I took copious screenshots on Dec 7th of Google Timeline so i'm not in total crap but it's highly affecting my life negatively.

**Comment by Freika** [comment]
@hopeseekr can  you please show me your queues in sidekiq?

**Comment by Beer17HWAM** [comment]
I imported my Google Timeline export in Owntracks and I imported the rec files from Owntracks in Dawarich. Importing the rec files is working fine. But the reverse Geocoding is taking a very long time. That is why I building my own photon docker for it. Tomorrow that thing is ready to go.

**Comment by hopeseekr** [comment]
@Beer17HWAM 

> I also have 12 year of Google history and it took me 2 weeks to reverse-geocode 2 years.

Man that's same with me. 12.5 days and only got from 2014 (with just 50 days in it, due to Android not exporting timeline data yet) through 2016 ... and it's slowing down precipitiously... and the amount enequeued has stayed the same??

It's goign to take over 3 months at this rate, and i'm not exaggerating. 

I lost 5% of my front brain associated with short-term memory storage in 2022 due to a flesh eating bacteria that migrated from my lip to my brain. I have nearly-zero spatial memory any more. I can get lost going 100 meters in a new area. I can't remember, literally, where i was 3 days ago, and Google Timeline is ESSENTIAL for my life. Google Timeline for Desktop, because Timeline for Android doens't have the "view for a month" feature at all...

I'm going to build my own parser that imports from the most recent to the oldest first. I doubt people need oldest data first and if it's going to take 3 months it's a huge frickin deal for me in my own personal life, since right now I can't remember or see past November 30th, which is really sucking.

I'm going to begin writing it tomorrow now that I understand the Points table now.  It'll be written in Rust (one of the fastest languages) and multithreaded doing 50-100 requests per CPU thread per second, if I can get a geoip decoder API that supports those levels of usage.

If I can't find one, I'm going to start literally supplicating these API providers as a goodwill endeavor just for this dawarich migration process for Google Timeline refugees, particularly those with spacial memory handicaps.

**Comment by hopeseekr** [comment]
Or here's an idea... My parser is going to insert all the lat/long coordinates first. I bet it can do that INSTANTLY, then have a batch process take the days it needs to do the geo lookup stuff... Hopefully Dawarich works with just lat/long coords...

Then who cares if takes 3 months if you get the basic data in an hour?

**Comment by hopeseekr** [comment]
> @hopeseekr can you please show me your queues in sidekiq?

![image](https://github.com/user-attachments/assets/63e514c0-2b47-4ef3-9ed1-350bed97d579)

That 1,721,000 number stays remarkably the same. It changed by -60,000 over 6 days. I've been watching it.

**Comment by Freika** [comment]
Okay, I will repeat this one last time, and then will just update the docs on the website and will be referring to it.

1. No matter how many hundreds of megabytes your location history takeout is, under most common circumstances it takes not more than 2 hours to import. And I'm talking about 2gb files.
2. What takes a lot longer time (due to reasons I highlighted repeatedly, including this very thread) is reverse geocoding process. HOWEVER. Reverse geocoding jobs in Sidekiq != importing jobs in Sidekiq. This is important to understand, because otherwise you're at risk of considering your importing process incomplete due to so many reverse geocoding jobs being unprocessed yet. If you, just like on the screenshot above, have an empty `imports` queue, you're done, your import is finished, either successfully or not. Everything you have in other queues have no business with imports. Now to the reverse geocoding.
3. With the previous two points in mind, it's important to understand, that the reverse geocoding process is _completely optional_. You can happily use Dawarich with it being disabled, and you will still be able to look through your location history on the map, navigate by days, months and years and so on. What you'll be missing on, is the data on the address, city and country (along with some other data) of where each and every point was recorded.

Now, I hope I made it clear.

So, @hopeseekr:

>Or here's an idea... My parser is going to insert all the lat/long coordinates first. I bet it can do that INSTANTLY, then have a batch process take the days it needs to do the geo lookup stuff... Hopefully Dawarich works with just lat/long coords...

> Then who cares if takes 3 months if you get the basic data in an hour?

Dawarich does EXACTLY what you're asking about in this message. Already, since, probably, day one. Well okay, maybe month one.

In short: importing process is not supposed to take more than two weeks and it actually does not take that long.

@Beer17HWAM I saw your ticket and I'll be ready to answer when I have the time and willpower, there is no need to highlight it in other tickets, thank you.

**Comment by hopeseekr** [comment]
@Freika Your app is **broken** if it is supposed to import immediately.

Fact is, I couldn't figure out what day I entered the UAE until yesterday, because it took **35 days** to import the data to the point that it showed up in the web interface.

It finished running today at January 2nd at 00:04:32 UTC.
